File: app/src/main/java/com/u1fukui/bbs/customview/RecyclerViewScrolledEndSubject.java
package com.u1fukui.bbs.customview;

import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;

public class RecyclerViewScrolledEndSubject {

    enum Irrelevant {
        INSTANCE
    }

    private static final int THRESHOLD_COUNT = 5;

    private RecyclerView recyclerView;

    private Subject subject = PublishSubject.create();

    private RecyclerView.OnScrollListener listener = new RecyclerView.OnScrollListener() {
        @Override
        public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
            LinearLayoutManager manager = (LinearLayoutManager) recyclerView.getLayoutManager();

            int visibleCount = recyclerView.getChildCount();
            int totalCount = manager.getItemCount();
            int firstPos = manager.findFirstVisibleItemPosition();

            if (totalCount <= firstPos + visibleCount + THRESHOLD_COUNT) {
                subject.onNext(Irrelevant.INSTANCE);
            }
        }
    };

    public RecyclerViewScrolledEndSubject(RecyclerView recyclerView) {
        if (!(recyclerView.getLayoutManager() instanceof LinearLayoutManager)) {
            throw new IllegalArgumentException("RecyclerView needs to have LinearLayoutManager");
        }
        this.recyclerView = recyclerView;
        this.recyclerView.addOnScrollListener(listener);
    }

    public Subject connect() {
        return subject;
    }

    public void shutdown() {
        recyclerView.removeOnScrollListener(listener);
        recyclerView = null;
    }
}

File: app/src/main/java/com/u1fukui/bbs/customview/RecyclerViewScrolledEndSubject.kt
package com.u1fukui.bbs.customview

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView

import io.reactivex.subjects.PublishSubject
import io.reactivex.subjects.Subject

class RecyclerViewScrolledEndSubject(private var recyclerView: RecyclerView?) {

    private val subject = PublishSubject.create<Any>()

    private val listener = object : RecyclerView.OnScrollListener() {
        override fun onScrolled(recyclerView: RecyclerView?, dx: Int, dy: Int) {
            val manager = recyclerView!!.layoutManager as LinearLayoutManager

            val visibleCount = recyclerView.childCount
            val totalCount = manager.itemCount
            val firstPos = manager.findFirstVisibleItemPosition()

            if (totalCount <= firstPos + visibleCount + THRESHOLD_COUNT) {
                subject.onNext(Irrelevant.INSTANCE)
            }
        }
    }

    internal enum class Irrelevant {
        INSTANCE
    }

    init {
        if (recyclerView?.layoutManager !is LinearLayoutManager) {
            throw IllegalArgumentException("RecyclerView needs to have LinearLayoutManager")
        }
        this.recyclerView?.addOnScrollListener(listener)
    }

    fun connect(): Subject<*> {
        return subject
    }

    fun shutdown() {
        recyclerView?.removeOnScrollListener(listener)
        recyclerView = null
    }

    companion object {

        private const val THRESHOLD_COUNT = 5
    }
}
